sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,555 @@
+ created at `execution_scope::work_item` scope, with the specified coordination
+ scope.
+
+|===
+
+
+=== Launch queries [[launch]]
+
+An implementation's ability to satisify requests for specific forward progress
+guarantees for a kernel may be dependent upon: the kernel itself, and the
+features that it uses; the queue (and underlying device) to which the kernel is
+submitted; and the kernel's launch configuration.
+
+It is a user's responsibility to ensure that a kernel requesting specific
+forward progress guarantees uses a compatible launch configuration, using the
+`ext_oneapi_get_info` function and descriptors from the `info::kernel`
+namespace.
+
+[source,c++]
+----
+namespace sycl {
+
+class kernel {
+ public:
+  template <typename Param, typename... T>
+  typename Param::return_type ext_oneapi_get_info(T... args) const;
+};
+
+}
+----
+
+[source,c++]
+----
+template <typename Param, typename... T>
+typename Param::return_type ext_oneapi_get_info(T... args) const;
+----
+_Constraints_: Available only when the types `T...` described by the parameter
+pack match the type list defined by `Param::argument_types`.
+
+_Preconditions_: `Param` must be one of the `info::kernel` descriptors defined
+in this extension, and the type alias `Param::return_type` must be defined in
+accordance with the table below.
+
+_Returns_: Information about the kernel that applies when the kernel is
+submitted with the configuration described by the parameter pack `T...`.
+
+This extension adds several new queries to this interface, many of which
+already have equivalents in the `kernel_device_specific` or `device`
+namespaces.
+
+NOTE: Redefining these queries is necessary to define `Param::argument_types`.
+If this extension is proposed for a future SYCL standard, such duplication
+could be avoided.
+
+[source, c++]
+----
+namespace ext::oneapi::experimental::info::kernel {
+
+template <uint32_t Dimensions>
+struct max_work_item_sizes;
+
+struct max_work_group_size;
+struct max_num_work_groups; // also defined in sycl_ext_oneapi_root_group
+
+struct max_sub_group_size;
+struct max_num_sub_groups;
+
+}
+----
+
+[%header,cols="1,5,5,5"]
+|===
+|Kernel Descriptor
+|Argument Types
+|Return Type
+|Description
+
+|`template <uint32_t Dimensions>
+ max_work_item_sizes`
+|`sycl::queue`
+|`id<Dimensions>`
+|Returns the maximum number of work-items that are permitted in each dimension
+ of a work-group, when the kernel is submitted to the specified queue.
+
+|`max_work_group_size`
+|`sycl::queue`
+|`size_t`
+|Returns the maximum number of work-items that are permitted in a work-group,
+when the kernel is submitted to the specified queue.
+
+|`max_num_work_groups`
+|`sycl::queue`, `sycl::range`, `size_t`
+|`size_t`
+|Returns the maximum number of work-groups, when the kernel is submitted to the
+specified queue with the specified work-group size and the specified amount of
+work-group local memory (in bytes).
```

Comment:
I think we need a few additional words here to communicate that this is the maximum number of work-groups (that can execute with additional forward progress guarantees? that can safely communicate via atomics? that can synchronize via a global barrier?), and not some limit on the total global ND-range.

As a detail too, does "the specified amount of work-group local memory (in bytes)" count just the "dynamic" local memory set in the host code and passed via local accessors, or does it also count any "static" local memory declared in the kernel code itself using sycl_ext_oneapi_local_memory?

Reply:
> I think we need a few additional words here to communicate that this is the maximum number of work-groups (that can execute with additional forward progress guarantees? that can safely communicate via atomics? that can synchronize via a global barrier?), and not some limit on the total global ND-range.

Actually, this _could_ be a limit on the total global ND-range, because the query is with respect to whatever properties the kernel has.  If the kernel has a forward progress requirement, then the query implicitly means "number of work-groups that can execute with the requested forward progress guarantees"; if a kernel has no properties, then the query means "number of work-groups that can execute this kernel".  This has tripped a few people up, so I'll expand the description in the specification to clarify this.

> As a detail too, does "the specified amount of work-group local memory (in bytes)" count just the "dynamic" local memory set in the host code and passed via local accessors, or does it also count any "static" local memory declared in the kernel code itself using sycl_ext_oneapi_local_memory?

Dynamic only.  The amount of static local memory is already covered because it's part of the kernel definition.  Clarified in 7319e89db7c1.

Reply:
Interesting.  This is getting a little out in the weeds, but have we thought about what an implementation should return for `max_num_work_groups` for a kernel without the `use_root_sync` property?  AFAIK this query didn't exist in SYCL previously, and there were no (defined) maximums for the global ND-range.

Reply:
I hadn't thought about this before.

I think there have always been some implicit maximum values, because `nd_range` stores `id`, and `id` stores `size_t` -- even without considering device capabilities, we can have at most `std::numeric_limits<size_t>::max()` work-items in each dimension.  This query accepts a work-group size, so the maximum number of work-groups (on a device with "no limits") can be calculated by dividing the maximum value by the user-provided work-group size.

If an implementation has real limits, it can reflect those here too, but those values would be implementation-defined.

Reply:
Cool, I think any additional discussion is beyond the scope of this PR so I'm fine resolving this comment for now.  Thanks!